### PR TITLE
netcheck, tailcfg, interfaces, magicsock: survey UPnP, NAT-PMP, PCP

### PR DESCRIPTION
**Pull request by bradfitz**
Don't do anything with UPnP, NAT-PMP, PCP yet, but see how common they
are in the wild.

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/527)

### CODE REVIEW

File: net/netcheck/netcheck.go [hunk 1]
```diff
@@ -34,15 +37,26 @@ import (
...
-	RegionV4Latency       map[int]time.Duration // keyed by DERP Region ID
-	RegionV6Latency       map[int]time.Duration // keyed by DERP Region ID
+	UDP                   bool     // UDP works
+	IPv6                  bool     // IPv6 works
+	IPv4                  bool     // IPv4 works
+	MappingVariesByDestIP opt.Bool // for IPv4
+	HairPinning           opt.Bool // for IPv4
+
+	// UPnP is whether UPnP appears present on the LAN.
+	// Empty means not applicable.
```
Comment (danderson) [inline net/netcheck/netcheck.go:47]: s/not applicable/unknown/ maybe?

File: net/netcheck/netcheck.go [hunk 2]
```diff
@@ -599,6 +614,98 @@ func (rs *reportState) stopProbes() {
...
+			if mem.Contains(mem.B(res[:n]), mem.S(":InternetGatewayDevice:")) {
+				rs.setOptBool(&rs.report.UPnP, true)
+			}
+		case 5351:
+			if n == 12 && res[0] == 0x00 { // right length and version 0
+				rs.setOptBool(&rs.report.PMP, true)
+			}
+			if n == 60 && res[0] == 0x02 { // right length and version 2
+				rs.setOptBool(&rs.report.PCP, true)
+				// Delete the mapping.
```
Comment (danderson) [inline net/netcheck/netcheck.go:671]: Maybe a few more words that, unlike the other protocols, PCP has no way to not create a mapping, so we delete it explicitly?

Comment (danderson) [reply net/netcheck/netcheck.go:671]: I guess you could cheese it and set your local IP to something like 8.8.8.8 to force the PCP router to respond with an error? Responding with an error still proves that PCP exists.

File: net/netcheck/netcheck.go [hunk 3]
```diff
@@ -861,6 +972,11 @@ func (c *Client) logConciseReport(r *Report, dm *tailcfg.DERPMap) {
 		fmt.Fprintf(w, " v6=%v", r.IPv6)
 		fmt.Fprintf(w, " mapvarydest=%v", r.MappingVariesByDestIP)
 		fmt.Fprintf(w, " hair=%v", r.HairPinning)
+		if r.UPnP != "" || r.PMP != "" || r.PCP != "" {
```
Comment (danderson) [inline net/netcheck/netcheck.go:975]: Did this in a couple places. `hasPortmapServices()` helper on the report obj maybe?

File: net/netcheck/netcheck.go (hunk 2 above)
Comment (bradfitz) [reply net/netcheck/netcheck.go:671]: I accidentally set my IP wrong before and just never got a reply.

If I could reliably cause an error that'd be great but I'm not really keen on learning how to do that (and testing that it works).

File: net/netcheck/netcheck.go (hunk 3 above)
Comment (bradfitz) [reply net/netcheck/netcheck.go:975]: Done.

File: net/netcheck/netcheck.go (hunk 1 above)
Comment (bradfitz) [reply net/netcheck/netcheck.go:47]: went more explicit: not checked.

**Pull request merged**